adal/src/main/java/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -118,6 +118,78 @@ AuthenticationResult getAccessToken() throws AuthenticationException {
         return AuthenticationResult.createResult(accessTokenItem);
     }
 
+    AuthenticationResult getAccessTokenUsingAssertion() throws AuthenticationException {
+        final String methodName = ":getAccessTokenUsingAssertion";
+        final AuthenticationResult result = acquireTokenWithAssertion();
+        
+        if (result != null && !StringExtensions.isNullOrBlank(result.getAccessToken())) {
+            try {
+                mTokenCacheAccessor.updateTokenCache(mAuthRequest, result);
+            } catch (MalformedURLException e) {
+                Logger.v(TAG + methodName, "Access token fetched but unable to update token cache");
+                throw new AuthenticationException(ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_URL, e.getMessage(), e);
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Send token request with grant_type as assertion to token endpoint for getting new access token.
+     */
+    AuthenticationResult acquireTokenWithAssertion() throws AuthenticationException {
+        final String methodName = ":acquireTokenWithAssertion";
+        Logger.v(TAG + methodName, "Try to get new access token with the provided assertion.",
+                mAuthRequest.getLogInfo(), null);
+
+        // Check if network is available, if not throw exception. 
+        HttpUtil.throwIfNetworkNotAvailable(mContext);
+
+        final AuthenticationResult result;
+        final String samlAssertion = mAuthRequest.getSamlAssertion();
+        final String assertionType = mAuthRequest.getAssertionType();
+
+        try {
+            final JWSBuilder jwsBuilder = new JWSBuilder();
+            final Oauth2 oauthRequest = new Oauth2(mAuthRequest, mWebRequestHandler, jwsBuilder);
+
+            result = oauthRequest.refreshTokenUsingAssertion(samlAssertion, assertionType);
+            if (result != null && StringExtensions.isNullOrBlank(result.getRefreshToken())) {
+                Logger.i(TAG + methodName, "Refresh token is not returned or empty", "");
+                // we have reached this point because we couldnt find the refresh token/use it
+                // so we cant set the refresh token
+            }
+        } catch (final ServerRespondingWithRetryableException exc) {
+
+            Logger.e(TAG + methodName,
+                    "Error in assertion for request. ",
+                    "Request: " + mAuthRequest.getLogInfo()
+                            + " " + ExceptionExtensions.getExceptionMessage(exc)
+                            + " " + Log.getStackTraceString(exc),
+                    ADALError.AUTH_FAILED_NO_TOKEN,
+                    null);
+
+            throw new AuthenticationException(
+                    ADALError.AUTH_FAILED_NO_TOKEN, ExceptionExtensions.getExceptionMessage(exc),
+                    new AuthenticationException(ADALError.SERVER_ERROR, exc.getMessage(), exc));
+        } catch (final IOException | AuthenticationException exc) {
```

Comment:
The `catch` block for the `IOExcepton`, `AuthenticationException` appears to contain the same logic as `ServerRespondingWithRetryableException`; is there any reason we should not collapse these into the same `catch` block?

Reply:
Yeah, there was some other code there(which was distinguishing the two catch blocks) initially, which was not needed later. Now these 2 catch blocks are doing the same thing, didnt realize that when I made changes. 

They can be collapsed into one catch block, I have changed it.

Reply:
> The `catch` block for the `IOExcepton`, `AuthenticationException` appears to contain the same logic as `ServerRespondingWithRetryableException`; 

Yeah, there was some other code there(which was distinguishing the two catch blocks) initially, which was not needed later. Now these 2 catch blocks are doing the same thing, didnt realize that when I made changes.

> is there any reason we should not collapse these into the same `catch` block?

They can be collapsed into one catch block, I have changed it.

Reply:
> is there any reason we should not collapse these into the same `catch` block?

The types in multi-catch are needed to be disjoint ServerRetryableException is a subclass of AuthenticationException.

Reply:
Thanks for clarifying.... So if `ServerRetryableException` is a subclass of `AuthenticationException` and these blocks contain the same logic can we remove the `ServerRespondingWithRetryableException` block entirely and still have the same behavior? Looks like yes, but perhaps I misunderstood

Reply:
Yeah you're right, I'll add the related changes and marking this as resolved.